nutils/solver.py

```diff
@@ -59,21 +59,20 @@
 
 argdict = types.frozendict[types.strictstr,types.frozenarray]
 
+def integral(arg):
+  return arg.as_evaluable_array()
+
 def integraltuple(arg):
-  if isinstance(arg, sample.Integral):
-    return arg,
-  for obj in arg:
-    if not isinstance(obj, sample.Integral):
-      raise TypeError('expected integral, got {}'.format(type(obj)))
-  return tuple(arg)
+  if hasattr(arg, 'as_evaluable_array'):
```

Comment:
In theory: yes. But this requires `typing.Protocol`, which is only available since 3.8, or adding `typing_extensions` as a dependency. I'm not sure if we want to go there. We could also revert the logic: if the argument is a tuple or a list do ... otherwise assume the argument satisfies `AsEvaluableArray`.

Reply:
Agreed let's not pull in `Protocol` yet. Inverting the logic might look a little more natural perhaps but I'm fine either way.